### PR TITLE
Preserve pkey->pmeth_engine on legacy keys

### PR DESCRIPTION
**Pull request by kcking**
Preserve pkey->pmeth_engine so that custom methods set on this key by an engine are preserved (for example in the libp11 pkcs11 engine).

This fixes an issue was introduced with https://github.com/openssl/openssl/pull/19965

### CODE REVIEW

File: crypto/engine/eng_pkey.c [hunk 1]
```diff
@@ -57,6 +57,7 @@ EVP_PKEY *ENGINE_load_private_key(ENGINE *e, const char *key_id,
                                   UI_METHOD *ui_method, void *callback_data)
 {
     EVP_PKEY *pkey;
+    bool copied_key;
```
Comment (beldmit) [inline crypto/engine/eng_pkey.c:60]: Test failures are relevant

`bool` type, `true`, and `false` are not supported in C89, AFAIR (and openssl doesn't correctly build with your patch). So could you please change it to some integer type?

Comment (kcking) [reply crypto/engine/eng_pkey.c]: Thank you, I changed it to an `int`.

**Comment by beldmit** [comment]
I confirm that your proposal doesn't break this test case: 
https://github.com/openssl/openssl/pull/19965#issuecomment-1382889805

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by beldmit** [comment]
@kcking does it fix #22945 ?

**Comment by t8m** [comment]
This looks like a terrible hack to me. Why doesn't the #22951 work?

**Comment by beldmit** [comment]
I'll try to dig it out on Thursday or Friday

**Comment by kcking** [comment]
https://github.com/openssl/openssl/pull/22951 fixes the issue that this PR tries to solve in a much cleaner way, thank you for looking into it t8m.

**Comment by beldmit** [comment]
Except it doesn't fix the use case I tried to fix initially

**Comment by t8m** [comment]
**OTC: How should we deal with regressions from https://github.com/openssl/openssl/pull/19965?**

My proposal is to simply revert that patch as the concrete problem that it solved in e_nss is now solved in e_nss code and it otherwise breaks more things that it solves such as #22508

**Comment by beldmit** [comment]
If e_nss has changed its behavior, the problematic code may be reverted

**Comment by beldmit** [comment]
I confirm that it works with e_nss 4.3

**Comment by petrovr** [comment]
> If e_nss has changed its behavior, the problematic code may be reverted

Hmm ....
NSS code does not trust on OpenSSL key material if is build with 3+ code base. So engine load method always re-creates key if build is with OpenSSL 3+.

As already pointed long time ago correction in OpenSSL code was not acceptable from my point of view. It does not resolve issue. Dunno how to explain difference between resolve vs mask an issue.

**Comment by t8m** [comment]
**OTC: This should not be applied and #23063 should be used instead.**

**Pull request closed without merging**